quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs

```diff
@@ -1284,9 +1294,110 @@ impl MetastoreService for PostgresqlMetastore {
+        let mut subresponses = Vec::with_capacity(request.subrequests.len());
+
+        let grouped_requests: HashMap<IndexUid, Vec<OpenShardsSubrequest>> = request
+            .subrequests
+            .into_iter()
+            .into_group_map_by(|subrequest| IndexUid::from(subrequest.index_uid.clone()));
+
+        for (index_uid, grouped_subrequests) in grouped_requests {
+            let index_metadata_request = IndexMetadataRequest {
+                index_id: None,
+                index_uid: Some(index_uid.to_string()),
+            };
+
+            let index_metadata: IndexMetadata = self
+                .index_metadata(index_metadata_request)
+                .await?
+                .deserialize_index_metadata()
+                .unwrap();
+
+            for subrequest in grouped_subrequests {
+                if !index_metadata.sources.contains_key(&subrequest.source_id) {
+                    return Err(MetastoreError::NotFound(EntityKind::Source {
+                        index_id: index_uid.index_id().to_string(),
+                        source_id: subrequest.source_id.to_string(),
+                    }));
+                }
+
+                let max_shard_opt: Option<PgShard> = sqlx::query_as::<_,PgShard>(
+                        r#"SELECT * FROM shards WHERE index_uid = $1 AND source_id = $2 ORDER BY shard_id DESC LIMIT 1;"#
+                    )
+                    .bind(index_uid.as_str())
+                    .bind(&subrequest.source_id)
+                    .fetch_optional(&self.connection_pool)
+                    .await
+                    .map_err(|error| {
+                        let index_uid: IndexUid = index_uid.clone();
+
+                        convert_sqlx_err(
+                            index_uid.index_id(),
+                            error,
+                        )
+                    })?;
+
+                if let Some(max_shard) = max_shard_opt {
+                    // The next_shard_id is +1 of the max_shard_id in the shards table.
+                    let max_shard_id = max_shard.shard_id + 1;
+                    let ok_next_shard_ids = max_shard_id - 1..=max_shard_id;
+
+                    if !ok_next_shard_ids.contains(&subrequest.next_shard_id) {
+                        warn!(
+                            "Control plane state is inconsistent with that of the metastore, \
+                             expected next shard ID `{}`, got `{}`",
+                            max_shard_id, subrequest.next_shard_id
+                        );
+                        return Err(MetastoreError::InconsistentControlPlaneState);
+                    }
+                }
+
+                let pg_shards: Vec<PgShard> = sqlx::query_as::<_, PgShard>(
+                        r#"
+                        INSERT INTO shards (index_uid, source_id, shard_id, leader_id, follower_id, shard_state, publish_position_inclusive, publish_token) 
```

Comment:
1. `RETURNING *` works only if there was modification to columns, on conflict it doesn't work and we end up with no output. It seems the only way to ensure that it always returns a value is to add a dummy update. This is a dumb solution and I am open to other queries to make this better.
2. I researched using transactions for this query so that we just get a single shard instead of a vector of all shards. The issue is the error `cannot insert multiple commands into a prepared statement` because a psql engine can only handle a single atomic command in a transaction. This is the simplest solution I could come up with.